!!! FEATURE: Overhaul node uri building

mhsdesign opened the pull request:
Resolves: https://github.com/neos/neos-development-collection/issues/4552

Introduces a new API for node-uribuilding to replace the `LinkingService`
Currently Neos9-Beta10 _does_ already contain a similar new `NodeUriBuilder` but this change overhauls the API to not expose the implementations of the Flow UriBuilder.

**Upgrade instructions**

The new `NodeUriBuilder` is not a singleton and thus cannot be injected but must be created via the `NodeUriBuilderFactory` with the current `ActionRequest` at hand.

Simple case:

```php
// before
#[Flow\Inject]
protected LinkingService $linkingService;

public function uriForNode(Node $node, ControllerContext $controllerContext): string
{
    return $this->linkingService
        ->createNodeUri(
            controllerContext: $controllerContext,
            node: $node
        );
}
```

```php
// after
#[Flow\Inject]
protected NodeUriBuilderFactory $nodeUriBuilderFactory;

public function uriForNode(Node $node, ActionRequest $actionRequest): UriInterface
{
    return $this->nodeUriBuilderFactory
        ->forActionRequest($actionRequest)
        ->uriFor(
            NodeAddress::fromNode($node)
        );
}
```

Advanced case:

```php
// before
#[Flow\Inject]
protected LinkingService $linkingService;

public function uriForNode(Node $node, ControllerContext $controllerContext): string
{
    return $this->linkingService
        ->createNodeUri(
            controllerContext: $controllerContext,
            node: $node,
            format: 'json',
            absolute: true,
            arguments: ['simple-query' => 'abc'],
            section: 'my-anchor'
        );
}
```

```php
// after
#[Flow\Inject]
protected NodeUriBuilderFactory $nodeUriBuilderFactory;

public function uriForNode(Node $node, ActionRequest $actionRequest): UriInterface
{
    $resolvedUri = $this->nodeUriBuilderFactory
        ->forActionRequest($actionRequest)
        ->uriFor(
            NodeAddress::fromNode($node),
            Options::create(forceAbsolute: true)->withCustomFormat('json')
        );

    return UriHelper::uriWithAdditionalQueryParameters($resolvedUri, ['simple-query' => 'abc'])
        ->withFragment('my-anchor');
}
```

Also notice that the PHP api for uri building will no longer keep the current format of the action request by default unlike in the Fusion api.
If this behaviour is required it should be implemented via `$options->withCustomFormat($actionRequest->getFormat());`

**Review instructions**

Currently in 9.0-dev the `NodeUriBuilder` is a not well though out abstraction. It has a really shallow implementation and doest abstract the uribuilder away, but rather exposes it to be set from outside. That makes it impossible to swap out the implementation and also the api is questionable. How do i really set the `format` option or add additional queryparameters?

With the overhaul of the api and lately new introduced `Shared\NodeAddress` we can go one step further and not use the legacy `FrontendRouting\NodeAdress` any further. The new address will fill its place and make thus also cross site linking will be possible: https://github.com/neos/neos-development-collection/issues/4441

We decided against adding the query parameters to be encoded as array option to the NodeUriBuilder `Options` and promote to use the helper `UriHelper::uriWithAdditionalQueryParameters`.
It is better to have it separated into a helper because its not core PSR standard behaviour (where the query is just a string) and anything further like `http_build_query` is php magic which does not obey an official standard.
Also another argument is that its just string manipulation _after_ the resolve was successful for the same reason a fragment cannot be provided as option but must be set via the PSR Api `withFragment`. 

Adjustment prs:
- [TASK: Adjust to introduction of Shared\NodeAddress and new NodeUriBuilder](https://github.com/neos/redirecthandler-neosadapter/pull/77)
- [TASK: adjust to overhauled node uri building](https://github.com/neos/neos-ui/pull/3802)

Related discussions:
- https://github.com/neos/flow-development-collection/pull/2744
- https://github.com/neos/flow-development-collection/issues/3354
- https://github.com/neos/flow-development-collection/pull/3316
- https://github.com/neos/neos-development-collection/issues/4564
- https://github.com/neos/neos-development-collection/issues/4564


**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
WIP: New `NodeUriBuilder` and `NodeUriSpecification`

mhsdesign commented on Neos.Neos/Classes/Fusion/ConvertUrisImplementation.php:
```diff
@@ -132,32 +139,36 @@ public function evaluate()
+                        $this->systemLogger->warning(sprintf('Could not resolve "%s" to a node uri.', $matches[0]), LogEnvironment::fromMethodName(__METHOD__));
                     }
                     $this->runtime->addCacheTag(
-                        CacheTag::forDynamicNodeAggregate($contentRepository->id, $nodeAddress->contentStreamId, NodeAggregateId::fromString($matches[2]))->value
+                        CacheTag::forDynamicNodeAggregate($nodeIdentity->contentRepositoryId, $nodeIdentity->workspaceName, $nodeIdentity->nodeAggregateId)->value
```
todo https://github.com/neos/neos-development-collection/issues/4810

mhsdesign committed:
WIP: Adjust to use new NodeAddress

mhsdesign committed:
FEATURE: Allow new nodeAddress to be serialized into uri

mhsdesign committed:
TASK: Use new NodeAddress in preview action

mhsdesign committed:
TASK: Use NodeAddress Uri serialization instead of json format for routePartHandler

mhsdesign committed:
TASK: Introduce NodeUri Options in favour of `NodeUriSpecification`

The api will be

```
public function uriFor(NodeAddress $nodeAddress, Options $options = null): UriInterface
```

mhsdesign commented:
todos now:
- [x] discuss flows base uri not being respected. Might be important for [such](https://github.com/neos/flow-development-collection/pull/3002#issuecomment-1503019924) trickery or for installing neos in a sub dir?
- [x] `FLOW_REWRITEURL` not respected see https://github.com/neos/flow-development-collection/pull/2744#issuecomment-1625414297
- [ ] test all usecases for resolving cross site (with base uri, without, blblablab
- [ ] then uribuilding failed, should the exception contain the arguments the route was tried to be build with?

planned followups:
- [ ] https://github.com/neos/neos-development-collection/issues/5069
- [ ] simplify node uri building (that no site detection result required) (e.g. make `public function resolve(array &$routeValues): bool` work by always using the cross site linking)
- [ ] readd RequestUriHostMiddleware, to set a `requestUriHost` for when resolving
- [ ] the frontend routing tests accept a `contentStreamId` as parameter but we work on workspaces and we only work on live, thus we can change it to
```diff
- the matched node should be "carl-destinode" in content stream "cs-identifier" and dimension '{"market":"DE", "language":"en"}'
+ the matched node should be "carl-destinode" in workspace "live" and dimension '{"market":"DE", "language":"en"}'
```
same for `The URL "/david-nodenborough" should match the node "sir-david-nodenborough" in content stream "cs-identifier" and dimension "{}"` while also here specifying that its `workspace "live"` is needless but we might as well keep it.
- [ ] the uribuilding currently needs the `request` set in fusion, but what is the fallback? That is a seperate issue but we _could_ make it a feature to pass the `baseUri` along in fusion
 -> add `NodeUriBuilderFactory::forBaseUri`
- [ ] `\Neos\Neos\Fusion\Helper\LinkHelper::resolveNodeUri` was changed in Neos 9 and now contains this `$controllerContext` as param ... that should be adjusted

bwaidelich commented on Neos.ContentRepository.Core/Classes/DimensionSpace/DimensionSpacePoint.php:
```diff
@@ -93,6 +93,11 @@ final public static function fromUriRepresentation(string $encoded): self
 
+    final public function toUriRepresentation(): string
+    {
+        return base64_encode(json_encode($this->coordinates, JSON_THROW_ON_ERROR));
```
The IDE will complain that this is missing a `@throws JsonException` and I agree to it – or rather: We should never throw JsonExceptions from core domain code IMO.

For the DSP I would suggest to introduce `public function toJson(): string`, catch any JsonException in there and rethrow as RuntimeException and then use it here and everywhere else.

Same for `fromUriRepresentation`.

(could be done in a separate PR ofc)

mhsdesign replied:
Knowing this is currently an already big enough of a change i didnt want to introduce anything new and seeing `fromUriRepresentation` already existed this was just something that was missed.

Lets do it in a followup: https://github.com/neos/neos-development-collection/issues/5072

bwaidelich commented on Neos.Neos/Classes/Fusion/NodeUriImplementation.php:
```diff
@@ -160,32 +160,39 @@ public function evaluate()
+                    forceAbsolute: $this->isAbsolute(),
+                    format: $this->getFormat(),
+                    routingArguments: $this->getAdditionalParams()
+                )
```
nitpick: large try/catch blocks always make me a bit suspicious even though it's not a real problem here.
What do you think of a construct like:
```php
$options =  Options::create(
    forceAbsolute: $this->isAbsolute(),
    format: $this->getFormat(),
    routingArguments: $this->getAdditionalParams()
 );
try {
    $resolvedUri = $nodeUriBuilder->uriFor(NodeAddress::fromNode($node), $options));
} catch (NoMatchingRouteException) {
    // ...
}
```

bwaidelich commented on Neos.Neos/Classes/Routing/NodeIdentityConverterAspect.php:
```diff
@@ -45,15 +46,13 @@ class NodeIdentityConverterAspect
      */
     public function convertNodeToContextPathForRouting(JoinPointInterface $joinPoint): array|string
     {
+        // todo whyyyy does `Route::resolves` even call `convertObjectsToIdentityArrays`? And why nodes?
```
```suggestion
        // todo whyyyy does `Route::resolves` even call `convertObjectsToIdentityArrays`? And why nodes?
```
What's the TODO here?

mhsdesign replied:
yeah ill remove them again, maybe you could explain that thing to me and if we can remove it :D 
-> discussion: https://github.com/neos/neos-development-collection/issues/5069

bwaidelich commented on Neos.Neos/Classes/ViewHelpers/Link/NodeViewHelper.php:
```diff
@@ -317,30 +311,33 @@ public function render(): string
+        $nodeUriBuilder = $this->nodeUriBuilderFactory->forRequest($this->controllerContext->getRequest()->getHttpRequest());
+
+        // todo ->setAddQueryString($this->arguments['addQueryString'])
+        // todo ->setArgumentsToBeExcludedFromQueryString($this->arguments['argumentsToBeExcludedFromQueryString']);
```
what about those?

mhsdesign replied:
Concerning the todos in the fluid view helpers and fusion impl. this change REMOVES certain features as discussed with martin as they should have been actually deprecated (we missed that):

lets discuss this in this space i just opened: https://github.com/neos/neos-development-collection/issues/5076

once we have a conclusion we can proceed. Lucky case would be to just remove it as proposed.

mhsdesign replied:
pr to remove them which would allow us to move forward here: https://github.com/neos/neos-development-collection/pull/5079

bwaidelich commented on Neos.Neos/Classes/ViewHelpers/Link/NodeViewHelper.php:
```diff
@@ -150,6 +146,12 @@ class NodeViewHelper extends AbstractTagBasedViewHelper
 
+    /**
+     * @Flow\Inject
+     * @var NodeUriBuilderFactory
```
we mix annotations and attributes in this class -_-

mhsdesign replied:
yeah but its fluid 😎

bwaidelich commented on Neos.Neos/Classes/ViewHelpers/Uri/NodeViewHelper.php:
```diff
@@ -222,54 +225,50 @@ public function render(): string
+        $nodeUriBuilder = $this->nodeUriBuilderFactory->forRequest($this->controllerContext->getRequest()->getHttpRequest());
+
+        // todo ->setAddQueryString($this->arguments['addQueryString'])
+        // todo ->setArgumentsToBeExcludedFromQueryString($this->arguments['argumentsToBeExcludedFromQueryString']);
```
and those?

kitsunet commented on Neos.ContentRepository.Core/Classes/DimensionSpace/DimensionSpacePoint.php:
```diff
@@ -93,6 +93,11 @@ final public static function fromUriRepresentation(string $encoded): self
 
+    final public function toUriRepresentation(): string
+    {
+        return base64_encode(json_encode($this->coordinates, JSON_THROW_ON_ERROR));
```
I don't see the benefit of base64 here.

mhsdesign replied:
Yes this state is highly inefficient as concluded in https://github.com/neos/neos-development-collection/issues/5072#issuecomment-2119249935, but its nothing new but what came with the nodeAddress 6 years ago, so to shortcut the discussion i thought ill do it like this for now and followup with https://github.com/neos/neos-development-collection/pull/5067 where we can focus on this fully.

This pr is about 'bigger' things.

bwaidelich commented on Neos.Neos/Tests/Behavior/Features/Bootstrap/RoutingTrait.php:
```diff
@@ -261,6 +262,13 @@ public function theNodeShouldResolveToUrl(string $nodeAggregateId, string $conte
     public function theNodeShouldNotResolve(string $nodeAggregateId, string $contentStreamId, string $dimensionSpacePoint): void
     {
+        if (
+            ($this->getObject(ConfigurationManager::class)
+                ->getConfiguration(ConfigurationManager::CONFIGURATION_TYPE_SETTINGS, 'Neos.Flow.mvc.routes')['Neos.Flow'] ?? null) === true
+        ) {
```
That check is not exact enough I'm afraid, `Neos.Flow.mvc.routes` could be an array

mhsdesign replied:
? we access `'Neos.Flow.mvc.routes')['Neos.Flow']`

and that is a bool

bwaidelich replied:
But it _can_ be an array, for example:
```yaml
Neos:
  Flow:
    mvc:
      routes:
        'Neos.Flow':
          position: 'end'
```
Not highly likely, but definitely allowed. Why not check for !null?

mhsdesign replied:
yeah oke ... or rather !empty as it must not be `false` either

kitsunet commented:
Suggestion: https://github.com/neos/neos-development-collection/issues/5072

bwaidelich reviewed (commented):
Just some (partly nitpicky) comments upon first pass.
But I have a more important "concern"/question:
Putting URI-related serializers into the VO is a cross-cutting concern.
I think we should really add `DimensionSpacePoint::toJson()`  and `DimensionSpacePoint::fromJson()` but `toUriRepresentation` is weird – and it's currently only used within `NodeAddress`.

Instead I suggested to have a dedicated `NodeAddressSerializer` or so. It could even be static.


Also, this PR adds quite a lot of TODOs but with no concrete action, IMO we should agree on that first

mhsdesign commented on Neos.Neos/Classes/FrontendRouting/NodeUri/Options.php:
```diff
@@ -0,0 +1,64 @@
+ *
+ * @api for the factory methods; NOT for the inner state.
+ */
+final readonly class Options
```
naming this options will make it required to be aliased when we introduce also `Options` for flows uribuilder... but thats why its in a separate namespace `NodeUri` so one can use 

```php
use Neos\Neos\FrontendRouting\NodeUri;

    /**
      * @Flow\Inject
      * @var NodeUri\NodeUriBuilderFactory
      */
     protected $nodeUriBuilderFactory;

NodeUri \Options::create(forceAbsolute: true);
```

should be fine?

mhsdesign replied:
Aaaand `routingArguments` should probably be deprecated and get its own with'er?

> @deprecated if you meant to append query parameters,
> please use {@see \Neos\Flow\Http\UriHelper::withAdditionalQueryParameters} instead:

```php
UriHelper::withAdditionalQueryParameters($this->nodeUriBuilder->uriFor(...), ['q' => 'search term']);
```

mhsdesign replied:
on the other hand

```php
use Neos\Neos\FrontendRouting;

    /**
      * @Flow\Inject
      * @var FrontendRouting\NodeUriBuilderFactory
      */
     protected $nodeUriBuilderFactory;

FrontendRouting\Options::create(forceAbsolute: true);
```

might be even simpler

mhsdesign replied:
@kitsunet @bwaidelich i dont know about our name space philosophy but here the lattter / inlining it to `FrontendRouting` might make the most sense?

mhsdesign commented:
> Instead I suggested to have a dedicated `NodeAddressSerializer` or so. It could even be static.

so we basically have to decide between `NodeAddressSerializer::createFromUriString(string $encoded): NodeAddress` vs `NodeAddress::fromUriString` 

I have a slight preference for the latter, especially since we didn't attach it to the Node itself. That way the value object would have some real use and would not just be something passed along to yet another class. Christian complained that the code is not explorable, and that way we would put yet another abstraction in between that. We can also rename `NodeAddress` but to me always its main purpose was to be serialised at some point and having the capability - being written in php - to serialise into something uri conform and efficient.

kitsunet commented:
We have made the mistake many times to be super specific deep down with annoying consequences. toUriString is IMHO way too specific for something that is THE way to serialize a node for later retrieval. Queue Jobs, Session, other entities, Cookies IDK I see tonnes of potential use for a node address. Having it that specialised hampers that. Having it just json serializable and caring for the encoding on the usage/output side is really way better

bwaidelich commented:
> We can also rename NodeAddress but to me always its main purpose was to be serialised at some point

That's a good argument, but there are a few more details to agree upon:

- As @kitsunet stated, this is not only about URIs so we should rather call the methods something more generic like `serialize` and `unserialize` (e.g. currently u use it in the `EventSourcedFrontendNodeRoutePartHandler`: `return new MatchResult($nodeAddress->toUriString(), ...` but it's specifically _not_ about the URI representation in that case but rather the "contextPath 2.0")
- A single method for (de)serialization might not be sufficient because depending on the use case some of the parts are already defined. E.g. in the Neos backend we already know `contentRepositoryId`  and `workspaceName` most of the time – I think we discussed this before and agreed to duplicate that information in the query param but I can't recall for certain (and if we do that we should think about the behavior if `SiteDetectionResult` and serialized NodeAddress don't match)

All in all my personal favorite would be

```php
public static function fromJson(string $json): self {
    try {
        // ...
    } catch (\JsonException $e) {
        throw new \InvalidArgumentException(...);
    }
}

public static function unserialize(string $serialized): self {
    return self::fromJson(base64_decode($serialized));
}

public function toJson(): string {
    try {
        // ...
    } catch (\JsonException $e) {
        throw new \RuntimeException(...);
    }
}

public function jsonSerialize(): mixed {
    // ...
}

public function serialize(): string {
    return base64_encode($this->toJson());
}
```
for `DimensionSpacePoint` and `NodeAddress`: See [example](https://loilo.github.io/php-arena/#N4Igxg9gJgpiBcIA8B+ADgCzQHQHZ4DMBLXAQwBsACAJxlKgl3IE9KxzSBnTygESIC2MXJyKMAymlJgYABQgkALpUFpyMIbkU9s2AFKcJMakQpEAXqQBG6vMDzZFaEwDdSimJQIBXXGEViuJQA+sGQIorU3v4AFHiUCYlo3jZEYJSk1NSkrAAkkBDUUCTuMJzxiQCUlMAAvg5BSSnkaZSciu6tPn4BjF7UEAIAglk5MZnZeVDupJXwbTDkBDUViYm0it7UQbgwAO4LSzG50x2VANyrlPX4+E7Nre2d6d3+gf2DBowx7Sa4AOaUXIAK0MuDmh2W9jujkirGhjkRkRgm22lF2B04iwIMVBjGCsEgsGOeNwABpKJFvDAKQBWACMACYKXpxAB5ABywQAKgAJABKbIA6sFOcEAKL8wX8yoXByKRy1NjuMAYShxfRg8UADxkaF6QVyMGqCIVZsUGAGBwxlF0AElcG4WlARv9vJpFDq9Qafs4lDiAOQAMVIRHUUEpEEorM5lEJ0E8vxIgOwIAApOUQJGMnxBMJRBIpDJ5Ep5hmAxSQWDKzAALQAPn+KIAsmVOKQmzFKhT6QB2ekANgALAzGfSh7LLjDFDdFQ1HMlUukngEXr43n1fFiTGZzDAfpFk0Dt6YWnuoBCsUsVtONlsgleCPB4AQBgIvrgYlYuDBhwSYES+65Ceu4wBecp3LOdyLi0a49O8igQB+XbzEmAI3kicIYWajh3mipLBMIgHHBaRCcCy7JcnygoimKkrShBSJKmAKpqhqH5ejA+rvEaJryphloQNa+y2tg-K+AEQicdx3ycH6WiBiGYZgdmMYcpQREJrmmgFrgkjSHIChaGWnAVkCdaNi2bYdvu3aUH2g4jky9K0pO-EzvKUELg8cEbkEpLiMYp4WLZ8wTDk2G4Si95AqRnANgURQlB4nBTnOtzeUuXjrgaCw7meoVtIe6GmlFqJBN+WJ-ppxK5HFDaIchbmQXgUHEGQVC0PQjAsGwHDcJQHIJkMUBQLQA2qOoHo6JqRj5RY1i2LgpowV0OXvKE4S-NEihxNOq3LsVgL5IwHhaPyXEQKIiHUMwdpQGS7kHUVfzHXshQANZyQZHKkEIj37T52n5oE+nFkZygnHmIig0WhlKADSLPWhx0dv8tD-KU93yiaXn3FlK5rfBfSvp8YIHq9QKkpe2KRYoWGlWaeEPtiz6k8MozMLiYL-sRVaMBSVI0pQo4UWK1HCqKXL0WyMqMWazGsequgcbqXG5bxdOwoJwkHPajpmC61Buh60k+nJfyKIpobhqplFxgBWko7a6aZtmpCDcNo3jZwJlmfz5LmQ2TaKK23A2V2Pb9sOo6Ds1TGeQ0azIx0q7ZcTQTsyMkzjJzQLhcwNPXtCaxrMz6IiY+e2lzX+ecwA2gG4RnYoF1oFdRA3XdUABgAugDtel7kBeN+91BfXDv1CH3A+D4k-A6bDBklsZL5vtnYzDw3AbFIvhbLxDffdlcg9b5MjdoxjWM9-3J9VFOax4ynzzp35lBbkFoEU0ewGfwVF6oWxPKRm5dHxszfMhSqv4hy8wTMcEC-944eQyvjWCr9cqNXJpeI6wD3IM3cmVGKBEapATimLKiApJZ0SlLLeWiplSKFVMrWauAzY8WNFremOsK56zEhJPMbDZLyStjEYMNsVKIWjPbEhntYAjTGm2P2NZg5WXDp2OyDkY7OV7Eg9KejoJA1eLlAKf8QooQyJzXBt5opohDsECAVhgQAUUMENw1BOAkQwGReWT9DHrT6Agsx2DXpWKROXKB1U-BwLql4+K9ZMHfCQTcKCJw5KUAALzAxhvvcGpY16DA3lzRuHAATeBsgGDJ9ZKABjKGZJuEAJK3QqekqpAZMhHynLkXAXsFEDUyTaIacjvZtlEbAAgpBvDkEUHUloLgYD+ygHJCkAZumwFrPSAMEEA4ZKBKsmA8ifYNSQlgzpgTzw7K6T0w59Yzm2U6ezD8OzBn7OGdwcBZNvgBy2R-ea5zMnPIOYo+APzgp7ngaY88EEAIYCjEgOF4o2TcgqGpeYAcKi3KgKijFFR2ZSM5PMYAuQHlgiOU1G4CQQWgUxTUXIlL-6kqwTcBF3IpwgDJCACZiFxK4AQOM8gWJ2WYDQAANWMLpBAIAAAcAA6AAzGykADT7ihwTBK7IewQC1CAA)

mhsdesign committed:
TASK: Serialize NodeAddress as json for uris

mhsdesign committed:
TASK: Fix preview uri building to ignore custom options

the custom options will likely be unintended as `uriFor` uses preview uris for non-live uris.
Such preview route will likely not exist, and thus it will just provoke hard to debug errors.

mhsdesign committed:
TASK: Document NodeUriBuilder and align behaviour to flows UriBuilder

- Use flows base uri if configured for sub dir installations
- Pass `uriPathPrefix` correctly along when base uri is configured
- support for `FLOW_REWRITEURLS === ß` is not planned as this option is to be removed

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into feature/overhaulNodeUriBuilding

mhsdesign committed:
TASK: Dont use `appendExceedingArguments` for preview action

mhsdesign committed:
TASK: Don't use options for `previewUriFor` but always build absolute uri

mhsdesign committed:
TASK: Fix WorkspacesController to redirect to node uris correctly

mhsdesign commented on Neos.Neos/Classes/FrontendRouting/NodeUri/NodeUriBuilderFactory.php:
```diff
@@ -0,0 +1,45 @@
+    /**
+     * @api
+     */
+    public function forRequest(ServerRequestInterface $request): NodeUriBuilder
```
regarding this method and the mandatory use of `getHttpRequest` in this case:

```php
$this->nodeUriBuilderFactory
    ->forRequest($actionRequest->getHttpRequest())
```

I have written quite a monolog regarding how it could be a nice thought to use the whole action request in itself to pass along as context and make getHttpRequest nullable in CLI context. This might sound controversy but i think its a good idea to think about it before introducing something like `ControllerContext2.0` or `UriBuildingContext`:

Discussion: https://github.com/neos/flow-development-collection/issues/3354

mhsdesign replied:
we decided for `forActionRequest` for now and might introduce a `forRouteDefaults` lateron

mhsdesign committed:
TASK: Refine and document format and routingArguments node uri building options

mhsdesign committed:
TASK: Document when `NoMatchingRouteException` will be thrown

mhsdesign committed:
TASK: Further improve comments and exception handling

mhsdesign commented on Neos.Neos/Classes/FrontendRouting/FrontendNodeRoutePartHandlerInterface.php:
```diff
@@ -20,8 +20,6 @@
  * Marker interface which can be used to replace the currently used FrontendNodeRoutePartHandler,
  * to e.g. use the one with localization support.
  *
- * TODO CORE MIGRATION
- *
```
what _can_ we migrate here anyway? The class name changed but extending it will render any custom code useless probably

mhsdesign commented on Neos.Neos/Classes/FrontendRouting/EventSourcedFrontendNodeRoutePartHandler.php:
```diff
@@ -261,15 +264,14 @@ public function resolveWithParameters(array &$routeValues, RouteParameters $para
-            // TODO log exception
+            $resolveResult = $this->resolveNodeAddress($nodeAddress, $currentRequestSiteDetectionResult->siteNodeName);
+        } catch (NodeNotFoundException | TargetSiteNotFoundException | InvalidShortcutException $exception) {
+            // TODO log exception ... yes todo
```
todo

mhsdesign committed:
TASK: Pass ActionRequest instead of destructuring HttpRequest

see https://github.com/neos/flow-development-collection/issues/3354

mhsdesign committed:
TASK: Place UriBuilder directly in `FrontendRouting`

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into feature/overhaulNodeUriBuilding

mhsdesign commented on Neos.Neos/Classes/FrontendRouting/Options.php:
```diff
@@ -0,0 +1,123 @@
+     * Note: The signature of this method might be extended in the future, so it should always be used with named arguments
+     * @see https://www.php.net/manual/en/functions.arguments.php#functions.named-arguments
+     */
+    public static function create(
+        bool $forceAbsolute = null
+    ): self {
+        return new self(
+            $forceAbsolute ?? false,
+            '',
+            []
+        );
+    }
+
+    /**
+     * Returns a new instance with the specified additional options
+     *
+     * Note: The signature of this method might be extended in the future, so it should always be used with named arguments
+     * @see https://www.php.net/manual/en/functions.arguments.php#functions.named-arguments
+     */
+    public function with(
+        bool $forceAbsolute = null
+    ): self {
```
these signatures dont feel quite right because they only have one argument ... and i wonder if we ever want to add any. Maybe rather

```
Options::forceAbsolute()
```

?

bwaidelich replied:
I agree. Or rather sth like `withEnforcedAbsoluteUri()` to be consistent

mhsdesign replied:
Went for 

```
Options::createForceAbsolute()
```

and 

```
Options::createEmpty()->withForceAbsolute();
```

felt the most aligned with `withCustomFormat` and `withCustomRoutingArguments`

mhsdesign commented on Neos.Neos/Classes/FrontendRouting/Options.php:
```diff
@@ -0,0 +1,123 @@
+     *
+     * See also {@link https://docs.neos.io/guide/rendering/rendering-special-formats}
+     */
+    public function withCustomFormat(string $format): self
```
i also noticed that the current format will NOT be takend from the `actionRequest` so one would need to take care of this themselves is this ok?

mhsdesign replied:
But i guess the Fusion helpers should default to the requests format? Either in php or better via

```php
format = ${request.format}
```

what do you think? ...

bwaidelich replied:
Good question, I'm not sure about that. When I played with HTMX, I used a custom format for the HX requests (see https://github.com/bwaidelich/Wwwision.HtmxTest/blob/main/Resources%2FPrivate%2FFusion%2FComponents%2FHtmxUri.fusion)
In the results I would _not_ want this format to be used for all links implicitly

kitsunet replied:
Currently it's the expected behavior, I have had it in both directions in projects. Sometimes I was happy it would automatically keep the xml I had in hte current request without me having to check it at every link, other times I would have wished to ignore it and use html by default 🤷

mhsdesign replied:
yeah i guess for fusion (especially as the magic is currently still part of the `Neos.Fusion:ActionUri`) i think it would make sense to keep it for the `Neos.Neos:NodeUri` as well.

~Ill vote for making it a default value on fusion side so its directly visible to the human eye? or should it be part of the fusion php implementation?~

edit on the other hand this is core behaviour and has to be documented via comment. Otherwise it would be weird that we access the `request` differently in two places (once for the uri builder and then for the format)

bwaidelich replied:
My gut feeling is still: html is like the 99% case and I would want to explicitly change that if I wanted to create links to other resources in the same format (e.g. in a JSON response I _might_ want to link to other JSON resources, but I would expect the node URI to still have the default format:
```json
{
  "title": "foo",
  "contactUri": "https://domain.tld/contact",
  "includedResources": {
    "contacts": [
      {"id": "123", "href": "/api/contacts/123.json"}
    ]
  }
}
```

Or maybe I can override the default format for the `RoutingDefaults` per controller/action?

mhsdesign replied:
I agree that the magic of keeping the format either has helped (by knowing it beforehand) or worse lead to confusion and i had to debug many times (because i forgot) why uri building failed ... the `@format: json` part in the exception dump `cat Data...` tends to go unnoticed for the untrained eye.
Still for this change it would be cruel to change behaviour like this ever so slightly and id prefer a dedicated discussion and separate pr.
I think we only have to agree that the PHP Node Uri building api doesnt do it by default and how to implement this magic again for the Fusion prototypes.

mficzel replied:
That makes sense ... to me. 

I would be fine with simply adding a default format of "html" to the fusion prototypes that is documented. 

That way you only have to read the fusion documentation to get it and not know about some magic behavior that works until you forget it and then causes trouble.

mhsdesign replied:
Im not against removing the format by request magic but id prefer a separate pr for that (and probably wont do it myself right now because i cant sell it as well as you do :D)
I guess this pr is a step in the right direction by removing that feature from the core node uribuilding and making it opt in from the outside. That means we can eventually get rid of it!

mhsdesign committed:
TASK: Restore feature to keep current format of action request when building uris from Fusion or Fluid

The magic of keeping the format either has helped (by knowing it beforehand) or worse lead to confusion and hard to debug uri building errors (the `@format: json` part in the exception dump tends to go unnoticed for the untrained eye)

Still for this change it would be cruel to change behaviour like this so we reimplement it.

mhsdesign committed:
TASK: Introduce dedicated `Options::withForceAbsolute`

The filter like signatures dont feel quite right because they only have one argument ... and its a yagni to add new args

kitsunet reviewed (approved):
Giving this 👀
Might do some more testing tomorrow, if it doesn't get merged.

mficzel commented on Neos.Neos/Configuration/Routes.Frontend.yaml:
```diff
@@ -7,7 +7,6 @@
   uriPattern:    'neos/preview'
   defaults:
     '@action':   'preview'
-  appendExceedingArguments: true
```
\o/ ... yeha that alone is worth it ... if only this would be the main fe route. I fear this needs `appendExceedingArguments` a while longer

mficzel reviewed (approved):
+1 By 👀 and 🧪 

Should be tested with: https://github.com/neos/neos-ui/pull/3802 as otherwise you end up with errors in the backend.

mhsdesign commented:
Thank you so much for all the reviews and discussions ;) Glad to have this merged!!! One step closer.
